File: app/src/main/java/fr/free/nrw/commons/wikidata/WikidataClient.java
package fr.free.nrw.commons.wikidata;

import com.google.gson.Gson;
import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;
import javax.inject.Singleton;

import fr.free.nrw.commons.wikidata.model.AddEditTagResponse;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import org.wikipedia.wikidata.Statement_partial;

@Singleton
public class WikidataClient {


    private final WikidataInterface wikidataInterface;
    private final Gson gson;

    @Inject
    public WikidataClient(WikidataInterface wikidataInterface, final Gson gson) {
        this.wikidataInterface = wikidataInterface;
        this.gson = gson;
    }

    /**
     * Create wikidata claim to add P18 value
     *
     * @return revisionID of the edit
     */
    Observable<Long> setClaim(Statement_partial claim, String tags) {
        return getCsrfToken()
            .flatMap(
                csrfToken -> wikidataInterface.postSetClaim(gson.toJson(claim), tags, csrfToken))
            .map(mwPostResponse -> mwPostResponse.getPageinfo().getLastrevid());
    }

    /**
     * Get csrf token for wikidata edit
     */
    @NotNull
    private Observable<String> getCsrfToken() {
        return wikidataInterface.getCsrfToken()
            .map(mwQueryResponse -> mwQueryResponse.query().csrfToken());
    }
}

File: app/src/main/java/fr/free/nrw/commons/wikidata/WikidataClient.kt
package fr.free.nrw.commons.wikidata

import com.google.gson.Gson
import fr.free.nrw.commons.wikidata.model.StatementPartial
import fr.free.nrw.commons.wikidata.model.WbCreateClaimResponse
import io.reactivex.Observable
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WikidataClient
    @Inject
    constructor(
        private val wikidataInterface: WikidataInterface,
        private val gson: Gson,
    ) {
        /**
         * Create wikidata claim to add P18 value
         *
         * @return revisionID of the edit
         */
        fun setClaim(
            claim: StatementPartial?,
            tags: String?,
        ): Observable<Long> =
            csrfToken()
                .flatMap { csrfToken: String? ->
                    wikidataInterface.postSetClaim(gson.toJson(claim), tags!!, csrfToken!!)
                }.map { mwPostResponse: WbCreateClaimResponse -> mwPostResponse.pageinfo.lastrevid }

        /**
         * Get csrf token for wikidata edit
         */
        private fun csrfToken(): Observable<String?> = wikidataInterface.getCsrfToken().map { it.query()?.csrfToken() }
    }
